spec/lib/pedicel/ec_spec.rb

```diff
@@ -0,0 +1,272 @@
+    end
+
+    it 'extracts the symmetric key using the private key and merchant id' do
+      expect(pedicel.symmetric_key(private_key: client.key, merchant_id: client.merchant_id)).to eq(symmetric_key)
+    end
+
+    it 'errs when missing parameters' do
+      expect{pedicel.decrypt(shared_secret: 'foo')}.to raise_error(ArgumentError)
+      expect{pedicel.decrypt(private_key: 'foo')}.to raise_error(ArgumentError)
+      expect{pedicel.decrypt(certificate: 'foo')}.to raise_error(ArgumentError)
+      expect{pedicel.decrypt(merchant_id: 'foo')}.to raise_error(ArgumentError)
+    end
+
+    it 'errs when both shared_secret and private_key is supplied' do
+      expect{pedicel.decrypt(shared_secret: 'foo', private_key: 'bar')}.to raise_error(ArgumentError)
+    end
+
+    it 'errs when both certificate and merchant_id is supplied' do
+      expect{pedicel.decrypt(certificate: 'foo', merchant_id: 'bar')}.to raise_error(ArgumentError)
+    end
+  end
+
+  describe '#shared_secret' do
+    let (:ss_and_epk) { @x ||= PedicelPay::Backend.generate_shared_secret_and_ephemeral_pubkey(recipient: client) }
+    let (:ss) { ss_and_epk.first }
+    let (:epk) { ss_and_epk.last }
+
+    let (:pedicel) do
+      backend.encrypt_and_sign(token, recipient: client, shared_secret: ss, ephemeral_pubkey: epk)
+
+      Pedicel::EC.new(token.to_hash)
+    end
+
+    it 'extracts the shared secret' do
+      expect(pedicel.shared_secret(private_key: client.key)).to eq(ss)
+    end
+
+    it "errs if the private key is from another curve than the token's ephemeral public key" do
+      key = OpenSSL::PKey::EC.new('wap-wsg-idm-ecid-wtls1') # Apple, do never switch to this curve.
+      key.generate_key
+
+      expect{pedicel.shared_secret(private_key: key)}.to raise_error(Pedicel::EcKeyError, /curve.*differ/)
+    end
+
+    it 'returns the wrong shared secret if the ephemeral public key is wrong' do
+      _, another_epk = PedicelPay::Backend.generate_shared_secret_and_ephemeral_pubkey(recipient: client)
+
+      expect(pedicel).to receive(:ephemeral_public_key).and_return(PedicelPay::Helper.ec_key_to_pkey_public_key(another_epk).to_pem)
+
+      expect(pedicel.shared_secret(private_key: client.key)).to_not eq(ss)
+    end
+
+    it 'errs if the ephemeral public key is malformatted' do
+      expect(pedicel).to receive(:ephemeral_public_key).and_return(Base64.decode64('An invalid ephemeral public key'))
+
+      expect{pedicel.shared_secret(private_key: client.key)}.to raise_error(Pedicel::EcKeyError, /invalid ephemeralPublicKey/)
+    end
+  end
+
+  describe '.symmetric_key' do
+    let (:merchant_id) { "\x15\xDA25#2\x99uV\x94\xC6?,\xB9$\x98\x97}\xD9MI\x98\xA2R\xCCGOM\xAA\xF2\x7F\x1C" }
+    let (:shared_secret) { "p)@\x89\x1E-\xA9\x8B`\x9C\xCE6((\xCB\x01\x8C\x17\x0F1\xE2?\xE1o\xA4\xEC&3\xCB\xEB9\xBF" }
+    let (:symmetric_key) { "\xE2[\x8Bu\e\x9B\xF0\\\x9ET\xBAe\xBCA\x9A\xAC\x8F\n\x1DE8\xBE\xF1U&?\xDD\x8A\x1A\xCA\x1E\xA9".b }
```

Comment:
Why only `#b` on the symmetric key string?

Reply:
To be honest, I don't recall. Could I ask you to poke at it (by adding/removing `#b`)?